docs-internal/generic-mapping.md

```diff
@@ -0,0 +1,249 @@
+# Generic Mapping Proposal
+
+Provide an API to map an arbitrary Java type to another arbitrary Java type.
+
+## Proposal
+
+The API consist of the following classes (may be changed due to implementation details):
+- `MapperManager` - the entry point to mapping of types
+- `Mapper` - a class capable of converting one type to another
+- `MapperProvider` - SPI class to support providers loaded through Java Service loader or configured through a builder
+- `MapperException` - `RuntimeException` thrown when a mapping is missing or the mapping itself failed
+
+This API should be also added to `Config` as an additional source of mappings.
+
+TODO an alternative name would be `Converter` instead of `Mapper`
+
+### Mapper Manager
+
+Mapping provides tools to map a type to another type.
+
+The types may be
+ - a java class (`String.class`)
+ - a generic type (`io.helidon.common.GenericType` for any java type, such as `Supplier<String>`)
+ 
+Mappings can be provided either through Java Service loader, or through
+ explicitly configured providers using a builder.
+ 
+The mapping function gets either a pair of `Class` objects, or a pair of `GenericType` objects
```

Comment:
Consider using just `java.lang.reflect.Type` instead.

Reply:
That is not very useful, as it does not carry generic information, so you could not define a function as
`T map(S object, Class<S> srcType, Class<T> targetType`
You would end up with
`Object map(Object source, Type srcType, Type targetType`

Reply:
For whatever it's worth, that is exactly what CDI, Guice, etc. and other parameterized-type-reflecting frameworks do when working with parameterized types in addition to just classes; you end up with almost what you indicate above, with the minor change that the return value is `T`: `T map(Object source, Type srcType, Type targetType)`.

Reply:
We  already use the `GenericType` in config, plan to use it in WebServer.
Also there are other specs, such as JAX-RS that uses its own `GenericType`
Also CDI is not using `Type`, but `TypeLiteral` that has the same function as our `GenericType`.
From `javax.enterprise.inject.Instance`: 
`<U extends T> Instance<U> select(TypeLiteral<U> subtype, Annotation... qualifiers);`